### PR TITLE
Allow search filter using regex

### PR DESCRIPTION
**Pull request by agarcia64**
Resolves issue https://github.com/nathan-weinberg/jeeves/issues/55

### CODE REVIEW

File: functions.py [hunk 1]
```diff
@@ -170,11 +170,35 @@ def get_jenkins_job_info(server, job_name):
 	return jenkins_api_info
 
 
-def get_jenkins_jobs(server, job_search_fields):
+def get_jenkins_jobs(server, job_search_fields, job_search_fields_regex=""):
```
Comment (amodi5) [inline functions.py:173]:
```suggestion
def get_jenkins_jobs(server, job_search_fields, job_search_fields_regex=None):
```

File: functions.py [hunk 2]
```diff
@@ -170,11 +170,35 @@ def get_jenkins_job_info(server, job_name):
...
 	''' takes in a Jenkins server object and job_search_fields string
 		returns list of jobs with given search field as part of their name
 	'''
 
+	# does job_search_fields_regex contain valid regex?
+	is_valid_regex = False
+
+	if (job_search_fields_regex is not None):
```
Comment (amodi5) [inline functions.py:181]:
```suggestion
	if (job_search_fields_regex):
```

File: functions.py [hunk 3]
```diff
@@ -170,11 +170,35 @@ def get_jenkins_job_info(server, job_name):
...
 
 
-def get_jenkins_jobs(server, job_search_fields):
+def get_jenkins_jobs(server, job_search_fields, job_search_fields_regex=""):
 	''' takes in a Jenkins server object and job_search_fields string
 		returns list of jobs with given search field as part of their name
 	'''
 
+	# does job_search_fields_regex contain valid regex?
+	is_valid_regex = False
```
Comment (amodi5) [inline functions.py:179]: Not needed, remove

File: functions.py [hunk 4]
```diff
@@ -170,11 +170,35 @@ def get_jenkins_job_info(server, job_name):
...
+
+	if (job_search_fields_regex is not None):
+		try:
+			re.compile(job_search_fields_regex)
+			is_valid_regex = True
```
Comment (amodi5) [inline functions.py:184]:
```suggestion
               relevant_jobs = []
	        # fetch all jobs from server that match the given regex
		all_jobs = server.get_job_info_regex(job_search_fields_regex)

		# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
		supported_versions = ['10', '13', '15', '16', '16.1']
		for job in all_jobs:
			job_name = job['name']
			if any(supported_version in job_name for supported_version in supported_versions):
				relevant_jobs.append(job)
				print(job_name)
```

File: functions.py [hunk 5]
```diff
@@ -170,11 +170,35 @@ def get_jenkins_job_info(server, job_name):
...
+		except re.error:
+			is_valid_regex = False
```
Comment (amodi5) [inline functions.py:186]:
```suggestion
			pass
```

File: functions.py [hunk 6]
```diff
@@ -170,11 +170,35 @@ def get_jenkins_job_info(server, job_name):
...
+		# fetch all jobs from server that match the given regex
+		all_jobs = server.get_job_info_regex(job_search_fields_regex)
+
+		# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
+		supported_versions = ['10', '13', '15', '16', '16.1']
+		for job in all_jobs:
+			job_name = job['name']
+			if any(supported_version in job_name for supported_version in supported_versions):
+				relevant_jobs.append(job)
+				print(job_name)
```
Comment (amodi5) [inline functions.py:200]: Remove

File: functions.py (hunk 1 above)
Comment (nathan-weinberg) [reply functions.py:173]: +1

File: functions.py (hunk 2 above)
Comment (nathan-weinberg) [reply functions.py:181]: Also remove parenthesis

File: functions.py (hunk 3 above)
Comment (nathan-weinberg) [reply functions.py:179]: +1

File: functions.py (hunk 2 above)
Comment (amodi5) [reply functions.py:181]: Also, make it consistent across all your change  (I had thought these parentheses were consistent across Jeeves)

File: functions.py [hunk 7]
```diff
@@ -170,10 +170,27 @@ def get_jenkins_job_info(server, job_name):
...
 
 
-def get_jenkins_jobs(server, job_search_fields):
+def get_jenkins_jobs(server, job_search_fields, job_search_fields_regex=None):
 	''' takes in a Jenkins server object and job_search_fields string
 		returns list of jobs with given search field as part of their name
 	'''
+	relevant_jobs = []
+
+	if (job_search_fields_regex):
```
Comment (amodi5) [inline functions.py:179]:
```suggestion
	if job_search_fields_regex:
```

Comment (nathan-weinberg) [reply functions.py:179]: +1

File: functions.py [hunk 8]
```diff
@@ -170,10 +170,27 @@ def get_jenkins_job_info(server, job_name):
...
+
+			# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
+			supported_versions = ['10', '13', '15', '16', '16.1']
+			for job in all_jobs:
+				job_name = job['name']
+				if any(supported_version in job_name for supported_version in supported_versions):
+					relevant_jobs.append(job)
+					print(job_name)
+		except re.error:
+			pass
```
Comment (amodi5) [inline functions.py:193]: Now that I think about it, @nathan-weinberg  do we wan't to show any error msg or something like "Sorry, Invalid Regex" to the user

Comment (nathan-weinberg) [reply functions.py:193]: Yes an error message such as below would be preferred

```suggestion
except re.error as e:
    print("Error compiling regex: {}".format(e))
```

Comment (nathan-weinberg) [reply functions.py:193]: Furthermore, if this try/except block is only meant to handle the compilation of the the regex, the rest of the code currently inside it should instead **follow** the block

@agarcia64 What is the expected behavior if the compilation fails? If it just continues as is currently implemented, won't the program just crash?

**Comment by nathan-weinberg** [comment]
@agarcia64 Changes will need to be made to `validate_config` as well - how do we expect this new functionality to behave? What if a user leaves both `job_search_fields` and `job_search_fields_regex` blank? What if they are both populated? What if one is and the other is not? 

All this needs to be defined in `README.md`

File: functions.py (hunk 8 above)
Comment (agarcia64) [reply functions.py:193]: As it is now, if the compilation fails it'll just pass and continue while ignoring the. regex. However, if the code inside that block is moved outside, then invalid regex will crash the program. 

I'll replace the pass with the error message!

**Comment by agarcia64** [comment]
> @agarcia64 Changes will need to be made to `validate_config` as well - how do we expect this new functionality to behave? What if a user leaves both `job_search_fields` and `job_search_fields_regex` blank? What if they are both populated? What if one is and the other is not?
> 
> All this needs to be defined in `README.md`

If both fields are populated, then the jeeves report will include the relevant jobs collected from both fields. I believe this should stay as is for users who want to utilize both fields.

As of now, the 'job_search_fields_regex' argument is optional and if left empty jeeves should operate as normal. The 'job_search_fields' argument is mandatory, as leaving it blank will cause an error due to the 'validate_config'. 

I think a good solution would be:
1. Allow both 'job_search_fields' and 'job_search_fields_reges' to be optional
2. If both fields are empty, throw an error within 'get_jenkins_jobs()'

This way, users who only want to search using regex can do so.

**Comment by amodi5** [comment]
> > @agarcia64 Changes will need to be made to `validate_config` as well - how do we expect this new functionality to behave? What if a user leaves both `job_search_fields` and `job_search_fields_regex` blank? What if they are both populated? What if one is and the other is not?
> > All this needs to be defined in `README.md`
> 
> If both fields are populated, then the jeeves report will include the relevant jobs collected from both fields. I believe this should stay as is for users who want to utilize both fields.
> 
> As of now, the 'job_search_fields_regex' argument is optional and if left empty jeeves should operate as normal. The 'job_search_fields' argument is mandatory, as leaving it blank will cause an error due to the 'validate_config'.
> 
> I think a good solution would be:
> 
> 1. Allow both 'job_search_fields' and 'job_search_fields_reges' to be optional
> 2. If both fields are empty, throw an error within 'get_jenkins_jobs()'
> 
> This way, users who only want to search using regex can do so.
> Do you think this would be a good idea?

Why not completely replace jobs_search_fields with the *_regex, how does Jenkins API behave differently when using get_jobs vs get_job_info_regex if correct regex is used

File: functions.py (hunk 8 above)
Comment (nathan-weinberg) [reply functions.py:193]: So if it ignores the regex, what happens next?

Comment (amodi5) [reply functions.py:193]: IMO we should just replace the fields config var with the regex var

Comment (agarcia64) [reply functions.py:193]: If the regex is ignored, it'll just move onto the pre-existing search using only job_search_fields. 

@amodi5 raises a good point, although I think it would be easier to keep the existing job_search_fields, delete the regex field and just check job_search_fields for valid regex. I tried doing this initially but couldn't get consistent results, however now that I've been working on this problem for a while I think I can get it done.

File: functions.py [hunk 9]
```diff
@@ -183,19 +184,23 @@ def get_jenkins_jobs(server, job_search_fields):
...
-		job_name = job['name']
-		if any(supported_version in job_name for supported_version in supported_versions):
-			for field in fields:
-				if field in job_name:
+	# check for fields that contain valid regex
+	for field in fields:
+		try:
+			# does the search field contain valid regex?
+			re.compile(field)
+			# fetch all jobs from server that match the given regex
```
Comment (nathan-weinberg) [inline functions.py:192]: Can we add a space between these lines?

File: functions.py [hunk 10]
```diff
@@ -183,19 +184,23 @@ def get_jenkins_jobs(server, job_search_fields):
...
+
+			# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
+			supported_versions = ['10', '13', '15', '16', '16.1']
+			for job in all_jobs:
+				job_name = job['name']
+				if any(supported_version in job_name for supported_version in supported_versions):
 					relevant_jobs.append(job)
-					break
+					print(job_name)
+		except re.error as e:
```
Comment (nathan-weinberg) [inline functions.py:202]: Add a space here as well

File: functions.py [hunk 11]
```diff
@@ -183,19 +184,23 @@ def get_jenkins_jobs(server, job_search_fields):
...
+			print("Error compiling regex: {}".format(e))
```
Comment (nathan-weinberg) [inline functions.py:204]: We may want a `sys.exit()` here as well; if Jeeves can't search jobs from Jenkins there's not really anywhere else to go from there as far as I see

Comment (nathan-weinberg) [reply functions.py:204]: Ah nevermind I see what's going on here; let me rephrase my question - if the given search field is not a regex (think one of our legacy search fields), will `re.compile` fail?

Comment (agarcia64) [reply functions.py:204]: No re.compile will not fail. If the given search does not contain valid regex, then the function server.get_job_info_regex(field) will still work by simply looking for jobs containing the string 'field'. 

Also, users may use both regex searches and normal searches at the same time, i.e.
job_search_fields: ^DFG-ceph-rhos, rgw$, unified
This will return:
1. All jobs starting with DFG-ceph-rhos
2. All jobs ending with rgw
3. All jobs containing the string unified

So the only time re.compile fails is when a string has invalid regex, i.e. job_search_fields: '[a-f', ^DFG-ceph-rhos
This will simply fail, output an error in the terminal and move onto '^DFG-ceph-rhos'

File: functions.py [hunk 12]
```diff
@@ -183,19 +184,26 @@ def get_jenkins_jobs(server, job_search_fields):
 	for i in range(fields_length):
 		fields[i] = fields[i].strip(' ')
 
-	# fetch all jobs from server
-	all_jobs = server.get_jobs()
+	# check for fields that contain valid regex
+	for field in fields:
+		print(field)
```
Comment (nathan-weinberg) [inline functions.py:189]: Remove

File: functions.py [hunk 13]
```diff
@@ -174,6 +174,7 @@ def get_jenkins_jobs(server, job_search_fields):
 	''' takes in a Jenkins server object and job_search_fields string
 		returns list of jobs with given search field as part of their name
 	'''
+	relevant_jobs = []
```
Comment (nathan-weinberg) [inline functions.py:177]: Can we put this back where it was originally?

File: functions.py [hunk 14]
```diff
@@ -183,19 +184,26 @@ def get_jenkins_jobs(server, job_search_fields):
...
+			supported_versions = ['10', '13', '15', '16', '16.1']
+			for job in all_jobs:
+				job_name = job['name']
+				if any(supported_version in job_name for supported_version in supported_versions):
 					relevant_jobs.append(job)
-					break
+					print(job_name)
+
+		except re.error as e:
+			print("Error compiling regex: {}".format(e))
```
Comment (nathan-weinberg) [inline functions.py:206]: I'm still not understanding the logic here - if `re.complie` throws an error, no jobs will appear in the report

Comment (agarcia64) [reply functions.py:206]: This isn't the case, I just tested this with the following parameters:
`job_search_fields: '[f, ^DFG-ceph-rhos'`

And it works as I think it should, here's the output:
                                        
Error compiling regex: unterminated character set at position 0 
DFG-ceph-rhos-10_director-rhel-virthost-3cont_1_to_2comp_1_to_3ceph-ipv4-vxlan-scale-up
DFG-ceph-rhos-10_director-rhel-virthost-3cont_2comp_3ceph-ipv4-vxlan-monolithic
DFG-ceph-rhos-10_director-rhel-virthost-3cont_3osdcompute-ipv4-vxlan-osdcompute-rgw
...

The report was sent to my email with the expected DFG-ceph-rhos jobs

Comment (nathan-weinberg) [reply functions.py]: Ack - I'll do some tests as well

**Comment by nathan-weinberg** [comment]
@amodi5 @jamepark4 @AlistairTonner Can y'all test this PR locally and see if it fits the needs for your respective DFGs?

File: functions.py [hunk 15]
```diff
@@ -183,19 +183,27 @@ def get_jenkins_jobs(server, job_search_fields):
...
+
+		try:
+			# does the search field contain a valid regex or search field?
+			re.compile(field)
+
+			# fetch all jobs from server that match the given regex or search
+			all_jobs = server.get_job_info_regex(field)
+
+			# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
+			supported_versions = ['10', '13', '15', '16', '16.1']
```
Comment (nathan-weinberg) [inline functions.py:198]: Move between L187 and L188

File: functions.py [hunk 16]
```diff
@@ -183,19 +183,27 @@ def get_jenkins_jobs(server, job_search_fields):
...
+			for job in all_jobs:
+				job_name = job['name']
+				if any(supported_version in job_name for supported_version in supported_versions):
 					relevant_jobs.append(job)
-					break
+					print(job_name)
+
+		except re.error as e:
+			print("Error compiling regex: {}".format(e))
```
Comment (nathan-weinberg) [inline functions.py:206]: Add something to this error log like "Skipping this field"

File: functions.py [hunk 17]
```diff
@@ -183,19 +183,27 @@ def get_jenkins_jobs(server, job_search_fields):
...
+			all_jobs = server.get_job_info_regex(field)
+
+			# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
+			supported_versions = ['10', '13', '15', '16', '16.1']
+			for job in all_jobs:
+				job_name = job['name']
+				if any(supported_version in job_name for supported_version in supported_versions):
 					relevant_jobs.append(job)
-					break
+					print(job_name)
```
Comment (nathan-weinberg) [inline functions.py:203]: Remove print

File: functions.py [hunk 18]
```diff
@@ -183,19 +183,27 @@ def get_jenkins_jobs(server, job_search_fields):
...
 	relevant_jobs = []
-	supported_versions = ['10', '13', '15', '16', '16.1']
-	for job in all_jobs:
-		job_name = job['name']
-		if any(supported_version in job_name for supported_version in supported_versions):
-			for field in fields:
-				if field in job_name:
+	for field in fields:
+
+		try:
```
Comment (nathan-weinberg) [inline functions.py:190]:
```suggestion
		try:

```

File: functions.py (hunk 16 above)
Comment (nathan-weinberg) [reply functions.py:206]:
```suggestion
			print("Error compiling regex: {} - skipping this search field...".format(e))
```

Comment (nathan-weinberg) [reply functions.py:206]: ^^ this please

File: functions.py (hunk 15 above)
Comment (nathan-weinberg) [reply functions.py:198]: No need to reinitialize supported_versions with each search field

File: functions.py (hunk 16 above)
Comment (agarcia64) [reply functions.py:206]: How does this look?
print("Error compiling regex, skipping this field: {}".format(e))

File: functions.py [hunk 19]
```diff
@@ -183,19 +183,26 @@ def get_jenkins_jobs(server, job_search_fields):
...
-			for field in fields:
-				if field in job_name:
+	for field in fields:
+		try:
+
+			# does the search field contain a valid regex or search field?
+			re.compile(field)
+
+			# fetch all jobs from server that match the given regex or search
+			all_jobs = server.get_job_info_regex(field)
```
Comment (amodi5) [inline functions.py:196]: @agarcia64 doesn't this give any errors from Jenkins API?

Comment (agarcia64) [reply functions.py:196]: If the regex is invalid (as in it is syntactically incorrect) then it will throw the exception error and move onto the next search field, if it exists.

Comment (amodi5) [reply functions.py:196]: So why can't we catch that exception?

Comment (agarcia64) [reply functions.py:196]: I'm not sure I know how to catch the exception. re.compile just compiles the given string into a regex object (that we don't save anywhere). I use that function because it'll will throw an error if given invalid regex.

Comment (amodi5) [reply functions.py:196]: I tested it and it gives the same error as re.error would give, ill suggest the code, pls test it

Error compiling regex: missing ), unterminated subpattern at position 5
No jobs found with given search field. Exiting...

File: functions.py [hunk 20]
```diff
@@ -183,19 +183,26 @@ def get_jenkins_jobs(server, job_search_fields):
...
-	for job in all_jobs:
-		job_name = job['name']
-		if any(supported_version in job_name for supported_version in supported_versions):
-			for field in fields:
-				if field in job_name:
+	for field in fields:
+		try:
+
+			# does the search field contain a valid regex or search field?
+			re.compile(field)
```
Comment (amodi5) [inline functions.py:193]:
```suggestion
			
```
remove

File: functions.py [hunk 21]
```diff
@@ -183,19 +183,26 @@ def get_jenkins_jobs(server, job_search_fields):
...
+			all_jobs = server.get_job_info_regex(field)
+
+			# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
+			for job in all_jobs:
+				job_name = job['name']
+				if any(supported_version in job_name for supported_version in supported_versions):
 					relevant_jobs.append(job)
-					break
+
+		except re.error as e:
```
Comment (amodi5) [inline functions.py:204]:
```suggestion
		except Exception as e:
```

File: functions.py (hunk 19 above)
Comment (agarcia64) [reply functions.py:196]: Ok, sounds good!
